### PR TITLE
Introduce ID normalizing manager

### PR DESCRIPTION
**Pull request by taylorsw04**
In preparation for eagerly allocating final IDs in IdCompressor, this PR introduces normalization logic (including support for eager finals) in a separate class that IdCompressor will later take a dependency on.

### CODE REVIEW

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+ */
+
+/* eslint-disable @typescript-eslint/restrict-plus-operands */
+
+import { assert, compareFiniteNumbers, compareFiniteNumbersReversed, fail, Mutable } from '../Common';
+import { FinalCompressedId, LocalCompressedId, SessionSpaceCompressedId } from '../Identifiers';
+import { AppendOnlyDoublySortedMap } from './AppendOnlySortedMap';
+
+/**
+ * Maps contiguous range of IDs created by a session between their local and final forms (i.e. normalization).
```
Comment (noencke) [inline experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:13]: test comment

Comment (noencke) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:13]: oh baby it works

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+				lastFinalRange[1] = lastFinal;
+			} else {
+				const lastLocalInRange = firstAlignedLocal - (lastAlignedFinal - firstAlignedFinal);
+				const alignedLocal = (lastLocalInRange - 1) as LocalCompressedId;
+				let rangeMap: FinalRangesMap<TRangeObject>;
+				if (isSingle) {
+					// Convert the single range to a range collection
+					rangeMap = SessionIdNormalizer.makeFinalRangesMap();
+					rangeMap.append(firstAlignedLocal, lastFinalRange);
+					finalRangesObj[1] = rangeMap;
```
Comment (noencke) [inline experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:267]: Mutating elements in a sorted list is a little scary, though fine if we know what we're doing of course. I wonder if we could `expensiveAssert` to ensure that the sort order is still valid after the mutation or something, just to give future code auditors some piece of mind without having to prove that the mutation is safe.

Alternatively, you could change your sorted map classes to allow `append` to overwrite the value for a given key iff that key is currently the last one in the list.

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+		});
+	}
+}
+
+/**
+ * Serialized table for normalizing IDs made by the local session.
+ *
+ * TODO: Move this into ID compressor persisted types when integrated.
+ */
+export interface SerializedSessionIdNormalizer {
```
Comment (noencke) [inline experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:385]: Doesn't need to be part of this PR, but it'd be cool to have a serialize and deserialize method on `AppendOnlySortedMap` itself. You could probably type it such that you only need to provide a `serializeKey` function if the key type is not a primitive, and only need to provide a `serializeValue` function if the value type is not a primitive.

**Review by noencke — CHANGES_REQUESTED** [review]
Reviewed SessionIdNormalizer.ts. Will do a separate review for the tests.

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+/**
+ * Maps contiguous range of IDs created by a session between their local and final forms (i.e. normalization).
+ * The local and final forms of IDs made by a session can be thought of as two equal-length sparse arrays, aligned such
+ * that normalizeLocalToFinal(locals[i]) === finals[i] and vice versa.
+ * Below is an example, with certain events marked with a ^ aligning with when they happened to illustrate how certain mappings can arise:
+ * Creation index:    0    1    2    3    4    5    6    7    8    9    10
+ *          Locals: [-1,  -2,   X,   X,   X,  -6,   X,  -8,   X,   X   -11]
+ *          Finals: [ 0,   1,   2,   3,   4,   10,  11,  12,  13,  14,   X]
+ *
+ *                    ^    ^two IDs allocated as locals since no cluster exists, new cluster created when acked
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:21]: Make it so.

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+				}
+			}
+		}
+		return undefined;
+	}
+
+	/**
+	 * Converts the local ID to it's corresponding final ID, if one exists.
+	 */
+	public getFinalId(localId: LocalCompressedId): [FinalCompressedId, TRangeObject] | undefined {
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:94]: Done, +test

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+		const localDelta = alignedLocal - localByIndex;
+		const finalId = firstFinal + localDelta;
+		if (finalId <= lastFinal) {
+			// `finalId` was an eagerly generated final ID
+			return finalId as SessionSpaceCompressedId;
+		}
+		return undefined;
+	}
+
+	private static extractFirstFinalFromRange<TRangeObject>(finalRange: FinalRange<TRangeObject>): FinalCompressedId {
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts]: Moved to module-level

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+		return finalRange[0];
+	}
+
+	private static makeFinalRangesMap<TRangeObject>(): FinalRangesMap<TRangeObject> {
+		return new AppendOnlyDoublySortedMap(
+			compareFiniteNumbersReversed,
+			SessionIdNormalizer.extractFirstFinalFromRange,
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts]: Fixed

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+	/**
+	 * Registers a local ID with this normalizer.
+	 * This must be the next ID in local space, meaning that it must be the -nth ID, where n is the number of new session-space IDs
+	 * the session has created. An example:
+	 *
+	 * Locals: [-1, -2,  X,  X]
+	 * Finals: [ 0,  1,  2,  3]
+	 * In this scenario, the only valid input to `addLocalId` would be -5.
+	 */
+	public addLocalId(localId: LocalCompressedId): void {
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts]: Made it no longer take a param, and returns the next localId instead. should allow cleaner usage. nice catch.

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+		const lastLocalRange = this.idRanges.last();
+		if (lastLocalRange !== undefined) {
+			const [firstLocal, [lastLocal, finalRanges]] = lastLocalRange;
+			if (localId === lastLocal - 1) {
+				// New local simply expands the last local range tracked
+				lastLocalRange[1][0] = localId;
+				return;
+			} else if (this.expensiveAsserts) {
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts]: Yup.

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts (hunk 2 above)
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:279]: Added a validity check.

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+				rangeMap.append(alignedLocal, [firstFinal, lastFinal, rangeObject]);
+				assert(alignedLocal >= lastLocal, 'Gaps in final space must align to a local.');
+			}
+		}
+	}
+
+	/**
+	 * Returns an enumerable of all session-space IDs known to this normalizer, in creation order.
+	 */
+	public *getAllIds(): IterableIterator<SessionSpaceCompressedId> {
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts]: done

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+				}
+			}
+			idRanges.append(firstLocal, [lastLocal, finalRanges]);
+		}
+		return normalizer;
+	}
+
+	public equals(
+		other: SessionIdNormalizer<TRangeObject>,
+		compareRangeObjects: (a: TRangeObject, b: TRangeObject) => boolean = () => true
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts]: whoops

File: experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts (hunk 3 above)
Comment (taylorsw04) [reply experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts:402]: agreed, for another PR :)

**Review by noencke — APPROVED** [review]

**Review by noencke — APPROVED** [review]

**Review by noencke — APPROVED** [review]

**Review by noencke — APPROVED** [review]

**Pull request merged**